samza-api/src/main/java/org/apache/samza/container/placements/ContainerPlacementMessage.java

```diff
@@ -0,0 +1,147 @@
+     */
+    SUCCEEDED,
+
+    /**
+     * Indicates that the container placement action is in failed
+     */
+    FAILED;
+  }
+
+  /**
+   * UUID attached to a message which helps in identifying duplicate request messages written to metastore and not
+   * retake actions even if metastore is eventually consistent
+   */
+  protected final UUID uuid;
+  /**
+   * Unique identifier for a deployment so messages can be invalidated across a job restarts
+   * for ex yarn bases cluster manager should set this to app attempt id
+   */
+  protected final String applicationId;
+  // Logical container Id 0, 1, 2
+  protected final String processorId;
+  // Destination host where container is desired to be moved
+  protected final String destinationHost;
+  // Optional request expiry which acts as a timeout for any resource request to cluster manager
+  protected final Optional<Duration> requestExpiryTimeout;
+  // Status of the current request
+  protected final StatusCode statusCode;
+
+  protected ContainerPlacementMessage(UUID uuid, String applicationId, String processorId, String destinationHost,
+      Duration requestExpiryTimeout, StatusCode statusCode) {
+    Preconditions.checkNotNull(uuid, "uuid cannot be null");
+    Preconditions.checkNotNull(applicationId, "applicationId cannot be null");
+    Preconditions.checkNotNull(processorId, "processorId cannot be null");
+    Preconditions.checkNotNull(destinationHost, "destinationHost cannot be null");
+    Preconditions.checkNotNull(statusCode, "statusCode cannot be null");
+    this.uuid = uuid;
+    this.applicationId = applicationId;
+    this.processorId = processorId;
+    this.destinationHost = destinationHost;
+    this.requestExpiryTimeout = Optional.ofNullable(requestExpiryTimeout);
+    this.statusCode = statusCode;
+  }
+
+  protected ContainerPlacementMessage(UUID uuid, String applicationId, String processorId, String destinationHost,
+      StatusCode statusCode) {
+    this(uuid, applicationId, processorId, destinationHost, null, statusCode);
+  }
+
+  public String getApplicationId() {
+    return applicationId;
+  }
+
+  public String getProcessorId() {
+    return processorId;
+  }
+
+  public String getDestinationHost() {
+    return destinationHost;
+  }
+
+  public StatusCode getStatusCode() {
+    return statusCode;
+  }
+
+  public Optional<Duration> getRequestExpiryTimeout() {
+    return requestExpiryTimeout;
+  }
+
+  public UUID getUuid() {
+    return uuid;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    ContainerPlacementMessage that = (ContainerPlacementMessage) o;
+    return uuid.equals(that.uuid) && getApplicationId().equals(that.getApplicationId()) && getProcessorId().equals(
+        that.getProcessorId()) && getDestinationHost().equals(that.getDestinationHost())
+        && getStatusCode() == that.getStatusCode();
```

Comment:
potential NPEs, we may use Objects.equals(uuid, that.uuid), which handles null checks.

Reply:
The constructor does not allow nulls, then I don't need this write